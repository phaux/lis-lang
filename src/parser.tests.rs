use super::*;

#[test]
fn prog_empty() {
    let ast = Parser::new("").parse_prog();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn prog_only_semis() {
    let ast = Parser::new(";;;;;").parse_prog();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn prog_just_vars() {
    let ast = Parser::new("a123; foo_bar; _foo; fooBar;").parse_prog();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn prog_just_1s() {
    let ast = Parser::new("1 1 1").parse_prog();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_let() {
    let ast = Parser::new("let x = 1;").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_print() {
    let ast = Parser::new("print 1").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_let_pat_obj() {
    let ast = Parser::new("let {x, y} = obj;").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_let_pat_obj_nested() {
    let ast = Parser::new("let {x: {y}} = obj;").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_let_pat_obj_rename() {
    let ast = Parser::new("let {x: y} = obj;").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_let_pat_obj_default() {
    let ast = Parser::new("let {x, y = 2} = obj;").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_let_pat_obj_rename_default() {
    let ast = Parser::new("let {x: y = 2} = obj;").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_ops() {
    let ast = Parser::new("1 + -2 * 3").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_ops_parens() {
    let ast = Parser::new("(1 + 2) * 3").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_unclosed_paren() {
    let ast = Parser::new("(1 + 2;").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_if() {
    let ast = Parser::new("if 1 then print 1").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_if_unclosed_curly() {
    let ast = Parser::new("if 1 then { print 1").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_if_else() {
    let ast = Parser::new("if x then { print 1 } else { print 0; }").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_if_else_chain() {
    let ast = Parser::new(
        "
        if x then print 1
        else if y then print 2
        else print 3
        ",
    )
    .parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_if_lone_else() {
    let ast = Parser::new("if 1 then a;b; else c;").parse_prog();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_print_str() {
    let ast = Parser::new(r#"print "hello\nworld""#).parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_str_concat() {
    let ast = Parser::new(r#""hello" ++ "world""#).parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_obj() {
    let ast = Parser::new("{ a: 1, b: 2 }").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_obj_expr() {
    let ast = Parser::new("{ x: 1 + 2, y: 3 * 4 }").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_prop_access() {
    let ast = Parser::new("obj.a.b.c").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_func_call() {
    let ast = Parser::new("func(a, b)").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_method_call() {
    let ast = Parser::new("obj.method()").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_func_call_missing_paren() {
    let ast = Parser::new("func(a, b").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_func_call_trailing_comma() {
    let ast = Parser::new("func(a, )").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_func_call_missing_comma() {
    let ast = Parser::new("func(a b)").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_func_call_just_comma() {
    let ast = Parser::new("func(,)").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_func_call_missing_arg() {
    let ast = Parser::new("func(,a)").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_func_call_prop_assign_bin_expr() {
    let ast = Parser::new("func().prop = 1 + 2").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_func_call_prop_access() {
    let ast = Parser::new("func().prop").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_obj_empty() {
    let ast = Parser::new("{}").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_obj_key_str() {
    let ast = Parser::new(r#"{ "name": "john", "age": 25 }"#).parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_obj_nested() {
    let ast = Parser::new(r#"{ user: { name: "john", age: 25 } }"#).parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_op_method_call() {
    let ast = Parser::new("(1 + 2).toString").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_func_decl() {
    let ast = Parser::new("fn add(a, b) { return }").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_return() {
    let ast = Parser::new("return 42").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_return_no_value() {
    let ast = Parser::new("return").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_func_decl_empty() {
    let ast = Parser::new("fn sayHello() { }").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_func_decl_missing_name() {
    let ast = Parser::new("fn (a, b) { a + b; }").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_func_decl_missing_params() {
    let ast = Parser::new("fn add { 1 + 2; }").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_func_decl_missing_body() {
    let ast = Parser::new("fn add(a, b)").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn stmt_func_decl_missing_brace() {
    let ast = Parser::new("fn add(a, b) { a + b").parse_stmt();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn comments_ignored_in_prog() {
    let src = r"
        // header
        let x = 1; /* mid */ let y = 2;
        print x // tail
        ;
    ";
    let ast = Parser::new(src).parse_prog();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn comments_inside_expr_and_args() {
    let src = r"
        func(
            1, // after first
            /* block */ 2
        ).prop /* between */ . toString ( /* none */ )
    ";
    let ast = Parser::new(src).parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn comments_only_prog() {
    let src = "// just a comment\n/* and a block */";
    let ast = Parser::new(src).parse_prog();
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn expr_comparison_operators() {
    let ast = Parser::new("a < b <= c > d >= e").parse_expr(0);
    insta::assert_debug_snapshot!(ast);
}
