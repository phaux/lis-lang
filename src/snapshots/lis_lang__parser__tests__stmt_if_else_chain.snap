---
source: src/parser.tests.rs
expression: ast
---
Ok(
    Span {
        offset: 9..78,
        pos: Pos {
            line: 1,
            col: 8,
        }..Pos {
            line: 3,
            col: 20,
        },
        node: If {
            keyword: Token {
                sigil: Keyword(
                    If,
                ),
                offset: 9..11,
                pos: Pos {
                    line: 1,
                    col: 8,
                }..Pos {
                    line: 1,
                    col: 10,
                },
            },
            condition: Span {
                offset: 12..13,
                pos: Pos {
                    line: 1,
                    col: 11,
                }..Pos {
                    line: 1,
                    col: 12,
                },
                node: Var {
                    name: "x",
                },
            },
            then_keyword: Token {
                sigil: Keyword(
                    Then,
                ),
                offset: 14..18,
                pos: Pos {
                    line: 1,
                    col: 13,
                }..Pos {
                    line: 1,
                    col: 17,
                },
            },
            cons_branch: Span {
                offset: 19..26,
                pos: Pos {
                    line: 1,
                    col: 18,
                }..Pos {
                    line: 1,
                    col: 25,
                },
                node: Print {
                    keyword: Token {
                        sigil: Keyword(
                            Print,
                        ),
                        offset: 19..24,
                        pos: Pos {
                            line: 1,
                            col: 18,
                        }..Pos {
                            line: 1,
                            col: 23,
                        },
                    },
                    expr: Span {
                        offset: 25..26,
                        pos: Pos {
                            line: 1,
                            col: 24,
                        }..Pos {
                            line: 1,
                            col: 25,
                        },
                        node: Num {
                            val: 1.0,
                        },
                    },
                },
            },
            else_keyword: Some(
                Token {
                    sigil: Keyword(
                        Else,
                    ),
                    offset: 35..39,
                    pos: Pos {
                        line: 2,
                        col: 8,
                    }..Pos {
                        line: 2,
                        col: 12,
                    },
                },
            ),
            alt_branch: Some(
                Span {
                    offset: 40..78,
                    pos: Pos {
                        line: 2,
                        col: 13,
                    }..Pos {
                        line: 3,
                        col: 20,
                    },
                    node: If {
                        keyword: Token {
                            sigil: Keyword(
                                If,
                            ),
                            offset: 40..42,
                            pos: Pos {
                                line: 2,
                                col: 13,
                            }..Pos {
                                line: 2,
                                col: 15,
                            },
                        },
                        condition: Span {
                            offset: 43..44,
                            pos: Pos {
                                line: 2,
                                col: 16,
                            }..Pos {
                                line: 2,
                                col: 17,
                            },
                            node: Var {
                                name: "y",
                            },
                        },
                        then_keyword: Token {
                            sigil: Keyword(
                                Then,
                            ),
                            offset: 45..49,
                            pos: Pos {
                                line: 2,
                                col: 18,
                            }..Pos {
                                line: 2,
                                col: 22,
                            },
                        },
                        cons_branch: Span {
                            offset: 50..57,
                            pos: Pos {
                                line: 2,
                                col: 23,
                            }..Pos {
                                line: 2,
                                col: 30,
                            },
                            node: Print {
                                keyword: Token {
                                    sigil: Keyword(
                                        Print,
                                    ),
                                    offset: 50..55,
                                    pos: Pos {
                                        line: 2,
                                        col: 23,
                                    }..Pos {
                                        line: 2,
                                        col: 28,
                                    },
                                },
                                expr: Span {
                                    offset: 56..57,
                                    pos: Pos {
                                        line: 2,
                                        col: 29,
                                    }..Pos {
                                        line: 2,
                                        col: 30,
                                    },
                                    node: Num {
                                        val: 2.0,
                                    },
                                },
                            },
                        },
                        else_keyword: Some(
                            Token {
                                sigil: Keyword(
                                    Else,
                                ),
                                offset: 66..70,
                                pos: Pos {
                                    line: 3,
                                    col: 8,
                                }..Pos {
                                    line: 3,
                                    col: 12,
                                },
                            },
                        ),
                        alt_branch: Some(
                            Span {
                                offset: 71..78,
                                pos: Pos {
                                    line: 3,
                                    col: 13,
                                }..Pos {
                                    line: 3,
                                    col: 20,
                                },
                                node: Print {
                                    keyword: Token {
                                        sigil: Keyword(
                                            Print,
                                        ),
                                        offset: 71..76,
                                        pos: Pos {
                                            line: 3,
                                            col: 13,
                                        }..Pos {
                                            line: 3,
                                            col: 18,
                                        },
                                    },
                                    expr: Span {
                                        offset: 77..78,
                                        pos: Pos {
                                            line: 3,
                                            col: 19,
                                        }..Pos {
                                            line: 3,
                                            col: 20,
                                        },
                                        node: Num {
                                            val: 3.0,
                                        },
                                    },
                                },
                            },
                        ),
                    },
                },
            ),
        },
    },
)
